coil-base/src/main/java/coil/request/ImageRequest.kt

```diff
@@ -0,0 +1,769 @@
+import coil.size.Precision
+import coil.size.Scale
+import coil.size.Size
+import coil.size.SizeResolver
+import coil.target.ImageViewTarget
+import coil.target.Target
+import coil.transform.Transformation
+import coil.transition.CrossfadeTransition
+import coil.transition.Transition
+import coil.util.EMPTY_DRAWABLE
+import coil.util.getDrawableCompat
+import coil.util.orEmpty
+import kotlinx.coroutines.CoroutineDispatcher
+import okhttp3.Headers
+import okhttp3.HttpUrl
+import java.io.File
+
+class ImageRequest private constructor(
+    val context: Context,
+
+    /** @see Builder.data */
+    val data: Any?,
+
+    /** @see Builder.key */
+    val key: String?,
+
+    /** @see Builder.target */
+    val target: Target?,
+
+    /** @see Builder.listener */
+    val listener: Listener?,
+
+    /** @see Builder.lifecycle */
+    val lifecycle: Lifecycle?,
+
+    /** @see Builder.dispatcher */
+    val dispatcher: CoroutineDispatcher?,
+
+    /** @see Builder.transition */
+    val transformations: List<Transformation>,
+
+    /** @see Builder.transition */
+    val transition: Transition?,
+
+    /** @see Builder.bitmapConfig */
+    val bitmapConfig: Bitmap.Config?,
+
+    /** @see Builder.colorSpace */
+    val colorSpace: ColorSpace?,
+
+    /** @see Builder.size */
+    val sizeResolver: SizeResolver?,
+
+    /** @see Builder.scale */
+    val scale: Scale?,
+
+    /** @see Builder.precision */
+    val precision: Precision?,
+
+    /** @see Builder.fetcher */
+    val fetcher: Pair<Class<*>, Fetcher<*>>?,
+
+    /** @see Builder.decoder */
+    val decoder: Decoder?,
+
+    /** @see Builder.allowHardware */
+    val allowHardware: Boolean?,
+
+    /** @see Builder.allowRgb565 */
+    val allowRgb565: Boolean?,
+
+    /** @see Builder.memoryCachePolicy */
+    val memoryCachePolicy: CachePolicy?,
+
+    /** @see Builder.diskCachePolicy */
+    val diskCachePolicy: CachePolicy?,
+
+    /** @see Builder.networkCachePolicy */
+    val networkCachePolicy: CachePolicy?,
+
+    /** @see Builder.headers */
+    val headers: Headers,
+
+    /** @see Builder.parameters */
+    val parameters: Parameters,
+
+    private val placeholderResId: Int,
+    private val placeholderDrawable: Drawable?,
+    private val errorResId: Int,
+    private val errorDrawable: Drawable?,
+    private val fallbackResId: Int,
+    private val fallbackDrawable: Drawable?
+) {
+
+    /** @see Builder.placeholder */
+    val placeholder: Drawable? get() = getDrawableCompat(placeholderDrawable, placeholderResId)
+
+    /** @see Builder.error */
+    val error: Drawable? get() = getDrawableCompat(errorDrawable, errorResId)
+
+    /** @see Builder.fallback */
+    val fallback: Drawable? get() = getDrawableCompat(fallbackDrawable, fallbackResId)
+
+    @JvmOverloads
+    fun newBuilder(context: Context = this.context) = Builder(this, context)
+
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        return other is ImageRequest &&
```

Comment:
Added a proper `equals`/`hashCode`/`toString` to begin addressing https://github.com/coil-kt/coil/issues/405. Still need to add equals/hashCode for a couple other classes.